sdk/core/azure-core/azure/core/pipeline/distributed_tracing/span.py

```diff
@@ -0,0 +1,190 @@
+import time
+from random import randint
+
+from typing import Any
+from typing_extensions import Protocol
+
+
+class AbstractSpan(Protocol):
+    @property
+    def span_id(self):
+        # type: () -> str
+        pass
+
+    def span(self, name="child_span"):
+        # type: (str) -> AbstractSpan
+        """Create a child span for the current span and append it to the child
+        spans list.
+
+        :type name: str
+
+        :param name: (Optional) The name of the child span.
+
+        :rtype: :class: `AbstractSpan`
+
+        :returns: A child Span to be added to the current span.
+
+        """
+        pass
+
+    def add_attribute(self, attribute_key, attribute_value):
+        # type: (str, str) -> None
+        """Add attribute to span.
+
+        :type attribute_key: str
+
+        :param attribute_key: Attribute key.
+
+
+
+        :type attribute_value:str
+
+        :param attribute_value: Attribute value.
+
+        """
+        pass
+
+    def add_annotation(self, description, **attrs):
+        # type: (str, Any) -> None
+        """Add an annotation to span.
+
+        :type description: str
+
+        :param description: A user-supplied message describing the event.
+
+                        The maximum length for the description is 256 bytes.
+
+
+        :type attrs: kwargs
+
+        :param attrs: keyworded arguments e.g. failed=True, name='Caching'
+
+        """
+        pass
+
+    def start(self):
+        # type: () -> None
+        """Set the start time for a span."""
+        pass
+
+    def finish(self):
+        # type: () -> None
+        """Set the end time for a span."""
+        pass
+
+    def to_header(self):
+        # type: () -> str
+        pass
+
+
+class AzureSpan():
```

Comment:
Do we really need a "generic" `AzureSpan` implementation? We are not trying to force the user to send context. We are only trying to make it really simple to enable doing so...

Reply:
@johanste can you clarify this, I had imagined that enabling distributed tracing was something that could be enabled by a global configuration, in which case any decorated function should create span automatically. I had also imagined, if tracing was not enabled via global configuration, that users could explicitly provide spans when they wanted granular tracing of a particular function. Is the first case not actually a requirement?

Reply:
I think that this gets back to the opencensus extension that Johan mentioned. I think that the extension will pass in a parent span explicitly. If other implementations allow that functionality we will accommodate that. If the client does not use an implementation or does not use the extension then I say we do nothing.

I think there is an alternate to the extension that I mention in the discussion above.

Reply:
0693978ee99741c3056c75da4b7b7e0c38a921f2